File: src/api/coupon/coupon.service.ts
import "server-only";

import type { ApiPageRequest, ApiResponse } from "@/api/shared/types";

import type { CouponDetail, CouponListItem, MemberCouponItem } from "../../feature/coupon/domain";
import type { CouponListQueryRequest } from "./coupon.dto";
import { couponRepository } from "./coupon.repository";

export const couponService = {
  // 쿠폰 목록 조회
  // 도메인 반환
  async getCoupons(query: CouponListQueryRequest, pageRequest: ApiPageRequest): Promise<ApiResponse<CouponListItem[]>> {
    const res = await couponRepository.getList(query, pageRequest);
    return {
      ...res,
      data: res.data?.map((item) => ({
        id: item.id,
        name: item.name,
        discountType: item.discountType,
        discountAmount: item.discountAmount,
        maxDiscountAmount: item.maxDiscountAmount,
        minOrderAmount: item.minOrderAmount,
        maxDiscountCount: item.maxDiscountCount,
        issueStartAt: item.issueStartAt,
        issueEndAt: item.issueEndAt,
        useStartAt: item.useStartAt,
        useEndAt: item.useEndAt,
        visible: item.visible,
      })),
    };
  },

  // 쿠폰 상세 조회
  // 도메인 반환
  async getCoupon(id: number): Promise<ApiResponse<CouponDetail>> {
    const res = await couponRepository.getDetail(id);
    if (!res.data) return { ...res, data: undefined };
    return {
      ...res,
      data: {
        id: res.data.id,
        name: res.data.name,
        description: res.data.description,
        discountType: res.data.discountType,
        discountAmount: res.data.discountAmount,
        maxDiscountAmount: res.data.maxDiscountAmount,
        minOrderAmount: res.data.minOrderAmount,
        maxDiscountCount: res.data.maxDiscountCount,
        issueStartAt: res.data.issueStartAt,
        issueEndAt: res.data.issueEndAt,
        useStartAt: res.data.useStartAt,
        useEndAt: res.data.useEndAt,
        visible: res.data.visible,
        createdAt: res.data.createdAt,
        updatedAt: res.data.updatedAt,
      },
    };
  },

  // 쿠폰 발급 현황 조회
  // 도메인 반환
  async getCouponIssues(couponId: number, pageRequest: ApiPageRequest): Promise<ApiResponse<MemberCouponItem[]>> {
    const res = await couponRepository.getIssues(couponId, pageRequest);
    return {
      ...res,
      data: res.data?.map((item) => ({
        id: item.id,
        memberId: item.memberId,
        used: item.used,
        usedAt: item.usedAt,
        expiredAt: item.expiredAt,
        issuedAt: item.issuedAt,
      })),
    };
  },
};

File: admin-api/src/main/java/com/tastyhouse/adminapi/coupon/CouponService.java
package com.tastyhouse.adminapi.coupon;

import java.time.LocalDateTime;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import com.tastyhouse.core.domain.coupon.domain.model.DiscountType;
import com.tastyhouse.core.domain.coupon.domain.vo.CouponId;
import com.tastyhouse.core.domain.member.domain.vo.MemberId;
import com.tastyhouse.core.domain.coupon.application.CouponCommandService;
import com.tastyhouse.core.domain.coupon.application.CouponQueryService;
import com.tastyhouse.core.domain.coupon.application.dto.CouponSearchCondition;
import com.tastyhouse.core.domain.coupon.application.dto.command.CouponCreateCommand;
import com.tastyhouse.core.domain.coupon.application.dto.command.CouponUpdateCommand;
import com.tastyhouse.core.domain.coupon.application.dto.result.CouponDetailResult;
import com.tastyhouse.core.domain.coupon.application.dto.result.CouponListItemResult;
import com.tastyhouse.core.domain.coupon.application.dto.result.MemberCouponItemResult;
import com.tastyhouse.core.shared.page.PageResult;
import com.tastyhouse.adminapi.common.PaginationResponse;
import com.tastyhouse.adminapi.coupon.response.CouponDetailResponse;
import com.tastyhouse.adminapi.coupon.response.CouponListItemResponse;
import com.tastyhouse.adminapi.coupon.response.MemberCouponAdminItemResponse;

@Service
@RequiredArgsConstructor
public class CouponService {

    private final CouponCommandService couponCommandService;
    private final CouponQueryService couponQueryService;

    public PaginationResponse<CouponListItemResponse> getCoupons(String name, String discountType, Boolean visible, int page, int size) {
        DiscountType type = discountType == null ? null : DiscountType.from(discountType);
        CouponSearchCondition condition = CouponSearchCondition.of(name, type, visible);
        PageResult<CouponListItemResponse> pageResult = couponQueryService.findAllCoupons(condition, page, size)
            .map(this::toCouponListItemResponse);
        return PaginationResponse.from(pageResult);
    }

    public Long createCoupon(
        String name,
        String description,
        String discountType,
        Integer discountAmount,
        Integer maxDiscountAmount,
        Integer minOrderAmount,
        Integer maxDiscountCount,
        LocalDateTime issueStartAt,
        LocalDateTime issueEndAt,
        LocalDateTime useStartAt,
        LocalDateTime useEndAt,
        boolean visible
    ) {
        CouponCreateCommand command = CouponCreateCommand.of(
            name, description, DiscountType.from(discountType), discountAmount, maxDiscountAmount,
            minOrderAmount, maxDiscountCount, issueStartAt, issueEndAt, useStartAt, useEndAt, visible
        );
        CouponId couponId = couponCommandService.createCoupon(command);
        return couponId.value();
    }

    public CouponDetailResponse getCoupon(Long id) {
        CouponId couponId = CouponId.of(id);
        CouponDetailResult couponDetail = couponQueryService.findDetailById(couponId);
        return toCouponDetailResponse(couponDetail);
    }

    public void updateCoupon(
        Long id,
        String name,
        String description,
        String discountType,
        Integer discountAmount,
        Integer maxDiscountAmount,
        Integer minOrderAmount,
        Integer maxDiscountCount,
        LocalDateTime issueStartAt,
        LocalDateTime issueEndAt,
        LocalDateTime useStartAt,
        LocalDateTime useEndAt,
        boolean visible
    ) {
        CouponId couponId = CouponId.of(id);
        CouponUpdateCommand command = CouponUpdateCommand.of(
            name, description, DiscountType.from(discountType), discountAmount, maxDiscountAmount,
            minOrderAmount, maxDiscountCount, issueStartAt, issueEndAt, useStartAt, useEndAt, visible
        );
        couponCommandService.updateCoupon(couponId, command);
    }

    public void deleteCoupon(Long id) {
        CouponId couponId = CouponId.of(id);
        couponCommandService.deleteCoupon(couponId);
    }

    public Long issueCoupon(Long couponId, Long memberId) {
        CouponId targetCouponId = CouponId.of(couponId);
        MemberId targetMemberId = MemberId.of(memberId);
        return couponCommandService.issueCouponByAdmin(targetCouponId, targetMemberId).value();
    }

    public PaginationResponse<MemberCouponAdminItemResponse> getIssuedCoupons(Long couponId, int page, int size) {
        PageResult<MemberCouponAdminItemResponse> pageResult = couponQueryService.findIssuedByCoupon(CouponId.of(couponId), page, size)
            .map(this::toMemberCouponAdminItemResponse);
        return PaginationResponse.from(pageResult);
    }

    private CouponListItemResponse toCouponListItemResponse(CouponListItemResult dto) {
        return CouponListItemResponse.from(
            dto.id(),
            dto.name(),
            dto.discountType().name(),
            dto.discountAmount(),
            dto.maxDiscountAmount(),
            dto.minOrderAmount(),
            dto.maxDiscountCount(),
            dto.issueStartAt(),
            dto.issueEndAt(),
            dto.useStartAt(),
            dto.useEndAt(),
            dto.visible()
        );
    }

    private CouponDetailResponse toCouponDetailResponse(CouponDetailResult dto) {
        return CouponDetailResponse.from(
            dto.couponId().value(),
            dto.name(),
            dto.description(),
            dto.discountType().name(),
            dto.discountAmount(),
            dto.maxDiscountAmount(),
            dto.minOrderAmount(),
            dto.maxDiscountCount(),
            dto.issueStartAt(),
            dto.issueEndAt(),
            dto.useStartAt(),
            dto.useEndAt(),
            dto.visible(),
            dto.createdAt(),
            dto.updatedAt()
        );
    }

    private MemberCouponAdminItemResponse toMemberCouponAdminItemResponse(MemberCouponItemResult dto) {
        return MemberCouponAdminItemResponse.from(
            dto.id(),
            dto.memberId().value(),
            dto.used(),
            dto.usedAt(),
            dto.expiredAt(),
            dto.createdAt()
        );
    }
}
